### DISCUSSION TITLE
Should a timestamp extension be included?

### DISCUSSION THREAD

**Discussion by stephenberry** [beve-org/beve#1] [Polls]
MessagePack has a [Timestamp extension](https://github.com/msgpack/msgpack/blob/8aa09e2a6a9180a49fc62ecfefe149f063cc5e4b/spec.md?plain=1#L420). Should EVE also include a timestamp as an extension? And, if so, how should it be formatted?

**Comment by SiebrenW** [comment]
Do we know the reason why MessagePack chose their format?
What I do see is that some of MessagePack's formats wil not hold up in the future, because Unix time 32-bit is running out and makes it effectively useless for future use within our lifetimes. You get another extension with their 34-bit stamp (another 200 years or so. No issue in our lifetime), but it's a far cry from 64-bit.
30-bit nanoseconds makes a lot of sense though. Don't need more bits, but it's a specific step where milliseconds could also make sense in 10 bits with the remaining 54-bits for unix-time. (or something)
I'm in favour of your own format and would advocate for at least 48-bit seconds from epoch at the very least. 32-bit is out, unless you want another epoch that can be configured.

**Reply by stephenberry** [reply]
@SirTates, thanks for your excellent thoughts. I'm in agreement. I like the idea of microseconds since 1 January 2000 using 64-bits.

**Reply by jcelerier** [reply]
Why not use a known standard and use PTP timestamps? 48 bits seconds, 32 bits nanoseconds, UNIX epoch (in TAI, not UTC)

**Comment by stephenberry** [comment]
@meftunca, In reference to your [Extensions Proposal](https://github.com/beve-org/beve-go/blob/main/docs/EXTENSION_PROPOSAL.md) I wanted to point out this original discussion on a time stamp format for BEVE.

Time stamps are critical for some of my use cases, and nanosecond accuracy is a must. I think PTP timestamps are a good idea. But, I also think there are a decent number of different time stamp formats that are useful in different cases.

The question is how do we play nice with JSON first and foremost, and then how can we also support MessagePack, PostgreSQL, and other timestamp use cases? I don't want the specification to be too complex, BEVE should be simple, but it should also be very useful.

I've pasted your proposal below. 

### 4 - Timestamp

High-precision timestamp with **optional timezone offset**. UTC assumed if timezone not specified.

**Design Philosophy**:
- ✅ **UTC by default**: If no timezone specified, UTC assumed (like MessagePack)
- ✅ **Optional timezone**: Can store offset when known (like CBOR Tag 0)
- ✅ **Compact**: Timezone only 2 bytes when needed
- ✅ **Interoperable**: Supports both PostgreSQL `TIMESTAMP` and `TIMESTAMPTZ` models
- ✅ **Best of both**: Flexibility + simplicity

**Layout**: `HEADER | PRECISION | EPOCH_SECONDS | SUB_SECOND [| TZ_OFFSET]`

**HEADER**: `0b00100'110` (1 byte)

**PRECISION** (1 byte):
```c++
Bit 0:   Has timezone offset (0=no/UTC, 1=yes)
Bits 1-3: Precision
  0 -> seconds only
  1 -> milliseconds (3 decimal places)
  2 -> microseconds (6 decimal places)
  3 -> nanoseconds  (9 decimal places)
```

**EPOCH_SECONDS**: Signed 64-bit integer (8 bytes, little-endian)
- Unix epoch (1970-01-01T00:00:00Z)
- Range: ~292 billion years before/after epoch

**SUB_SECOND**: Unsigned integer (precision-dependent)
```c++
milliseconds: uint16_t (2 bytes) -> 0-999
microseconds: uint32_t (4 bytes) -> 0-999,999
nanoseconds:  uint32_t (4 bytes) -> 0-999,999,999
```

**TZ_OFFSET**: Signed 16-bit integer (2 bytes, optional)
- Minutes offset from UTC (-1439 to +1439)
- Only present if precision bit 0 is set
- Examples: UTC+5:30 = +330, UTC-8:00 = -480, UTC = 0

**Total Size**:
- Seconds (UTC):      9 bytes (header + precision + epoch)
- Seconds (with TZ):  11 bytes (+ 2 bytes offset)
- Nanoseconds (UTC):  14 bytes
- Nanoseconds (TZ):   16 bytes (+ 2 bytes offset)

**JSON Representation** (RFC 3339):
```json
"2025-10-14T15:30:45.123456789Z"          // UTC (no timezone offset)
"2025-10-14T10:30:45.123456789-05:00"     // With timezone offset
```

**Timezone Handling**:
- **No timezone** (bit 0 = 0): Assumed UTC, like MessagePack
- **With timezone** (bit 0 = 1): Explicit offset stored, like CBOR Tag 0
- **Storage flexibility**: Application chooses based on data source
- **Display**: Application converts to local timezone when rendering
- **Example**: 
  - Server without timezone info → stores UTC (9-14 bytes)
  - User input with timezone → stores offset (11-16 bytes)
  - IoT device → always UTC (smaller payload)

**Comment by meftunca** [comment]
The suggestion I shared actually allows us to provide support while maintaining beve's speed and efficiency. I am about to finish support for all the extensions I suggested in our Golang port. In the next phase, I will create general usage scenarios with full integration tests under the e2e/ directory. I think the main areas we should focus on for beve are data accuracy, quick access, and typed features. Let's continue this discussion for timestamp and implement the most reasonable choice.

**Comment by meftunca** [comment]
I added my latest improvements to the Readme.md file. I'm also sharing the link to the final proposals.
[beve-extensions](https://github.com/beve-org/beve-go?tab=readme-ov-file#-beve-extensions-v130)
[Extensions proposals](https://github.com/beve-org/beve-go/blob/main/EXTENSIONS_README.md)


With the Golang version, you can test both the timestamp and the other type suggestions I have proposed.
